Feat/network pass

leogermani opened the pull request:
Creates the Network pass feature.

Also, this work surfaced a small issue with the plugin. The post_process_in_hub method was being called even for events that were triggered in the Hub itself. This would cause the local memberhips to get the post meta and they would become uneditable.

To be consistent, this method should not be called in such cases. Just as process_in_node is not called in node for their own events.

I slightly changed the architecture there and updated the docs. It should be good now.

** Testing

Setup

* On site A, create two membership plans
* In one of the plans, add any value to the new "Network ID" field

![image](https://github.com/Automattic/newspack-network/assets/971483/db54a04f-5be3-4c6a-9566-d9dbe7908f0e)


* Leave Network ID blank for the other membership plan
* Do the same on Site B, two memberships
* In one of the memberships, put the same Network ID

Now let's test the flow

* Create a new membership for a user in one of the sites - in the memberhips with a Network ID
* Confirm that a `newspack_network_woo_membership_updated` is triggered and added to the event log
* Once this event reaches the other site, confirm the user is created and a membership is granted to them in the membership plan that shares the same network id
* In the site where the membership was distributed to, confirm the link to edit the membership points to the original site, where the membership was first created
* Confirm the action links (when you hover the memberhip in the admin table) shows a message "Managed in {link}" with a link to the original membership

![image](https://github.com/Automattic/newspack-network/assets/971483/4c3131f1-6c1d-4f01-887d-f8bf97821999)

* Edit the orginal membership and change its status and confirm all status changes are properly propagated across the network
* Create a membership in the membership plan without a Network ID and confirm no events are triggered

leogermani committed:
feat: Add Network ID metabox to Membership plans

leogermani committed:
feat: add basic membership event handling

leogermani committed:
Merge branch 'trunk' into feat/network-pass

leogermani committed:
feat: redirect distributed membership edit link

leogermani committed:
chore: store meta key in a constant

leogermani committed:
chore: copy

adekbadek commented on includes/woocommerce-memberships/class-admin.php:
```diff
@@ -0,0 +1,149 @@
+		?>
+		<label for="newspack-network-id"><?php esc_html_e( 'Network ID', 'newspack-network' ); ?></label>
+		<input type="text" id="newspack-network-id" name="newspack_network_id" value="<?php echo esc_attr( $network_id ); ?>" />
+		<p><?php esc_html_e( 'If a Network ID is informed, the user memberships will be propagated to the network. The user will be granted with the membership with the matching Network ID in all other sites in the network.', 'newspack-network' ); ?></p>
```
```suggestion
		<p><?php esc_html_e( 'If a Network ID is set, the user memberships will be propagated on the network. Users will be granted the membership with the matching Network ID in all other sites in the network.', 'newspack-network' ); ?></p>
```

adekbadek replied:
@leogermani WDYT about these wording changes?

adekbadek commented on includes/woocommerce-memberships/class-admin.php:
```diff
@@ -0,0 +1,149 @@
+		if ( ! $managed_by_newspack_network ) {
+			return $actions;
+		}
+		$link     = get_edit_post_link( $post->ID );
```
The ID here is on the "receiving" site, not the site where the membership is managed, so the link will be wrong.

adekbadek commented on includes/woocommerce-memberships/class-admin.php:
```diff
@@ -0,0 +1,149 @@
+	 * @return array
+	 */
+	public static function post_row_actions( $actions, $post ) {
+		$managed_by_newspack_network = get_post_meta( $post->ID, self::NETWORK_MANAGED_META_KEY, true );
```
This will also be true on the membership-originating site, resulting in a confusing "Managed in <this-site>" message.

adekbadek commented on includes/woocommerce-memberships/class-admin.php:
```diff
@@ -0,0 +1,149 @@
+	 * @param int    $post_id The post ID.
+	 * @return string
+	 */
+	public static function get_edit_post_link( $link, $post_id ) {
```
The membership on the "receiving" site is still editable if visited directly (manually typing the post edit link). The post-save redirect goes to the wrong link, though (see comment on `post_row_actions` method). The membership updated this way (from the receiving site) is not updated on the origin site. This might be confusing. 

If the update is supposed to be one-way, the membership edit view should not be editable, just like post edit view of a distributed post.

leogermani committed:
fix: membership link and permission

leogermani committed:
fix: dont invoke process_in_hub for hub events

leogermani commented:
Thanks @adekbadek ,

I have fixed the link and changed the approach to block edition. I have now a check for the `edit_post` capability check. See what you think.

Also, this work surfaced a small issue with the plugin. The `post_process_in_hub` method was being called even for events that were triggered in the Hub itself. This would cause the local memberhips to get the post meta and they would become uneditable.

To be consistent, this method should not be called in such cases. Just as `process_in_node` is not called in node for their own events.

I slightly changed the architecture there and updated the docs. It should be good now.

(btw, I'm open to revisiting the method names)

leogermani committed:
fix: dont invoke process_in_hub for hub events

adekbadek committed:
Merge branch 'trunk' into feat/network-pass

# Conflicts:
#	includes/incoming-events/class-order-changed.php
#	includes/incoming-events/class-subscription-changed.php

adekbadek committed:
chore: lint fixes

adekbadek commented on includes/woocommerce-memberships/class-admin.php:
```diff
@@ -0,0 +1,187 @@
+class Admin {
+
+	/**
+	 * The membership plan custom post type.
```
```suggestion
	 * The membership plan custom post type (defined in the woocommerce-memberships plugin).
```

adekbadek reviewed (approved):
Pre-approving, pending some wording suggestions.

leogermani committed:
Update includes/woocommerce-memberships/class-admin.php

Co-authored-by: Adam Boro <aborowski24@gmail.com>

leogermani committed:
Update includes/woocommerce-memberships/class-admin.php

Co-authored-by: Adam Boro <aborowski24@gmail.com>

leogermani commented:
@adekbadek this is good to land. But leaving this un-merged to see how we want to proceed with the release, If we want to have this in an epic branch, a feature flag... or if it's fine since it will have no effect until a Network ID is set

adekbadek commented:
I think we're good to go, just remember to add public-facing docs after the release.

leogermani committed:
Merge branch 'trunk' into feat/network-pass
